sql/schemachanger: add back TXN_DROPPED, stop using OFFLINE

ajwerner opened the pull request:
The first two commits enable synthetic descriptors to be used during transaction
execution to influence the virtual tables.

We added the `OFFLINE` and performed write operations during drops to move the
descriptor into that state in order to deal with the fact that synthetically
dropped descriptors would not appear so in virtual tables. Fortunately, this
shortcoming has been rectified. We now revive the TXN_DROPPED state. This
proves to be important because the `OFFLINE` state is not backwards compatible,
and the `TXN_DROPPED` state is. Ideally we'd find a way to move the descriptors
straight to `DROPPED` in the pre-commit case for all the cases we today support,
but doing that would require an overhaul of the dependency rules which, at this
point in the cycle, feels risky.

Fixes https://github.com/cockroachdb/cockroach/issues/86626
Fixes #77148
Fixes #86691

Release justification: Fixes a release-blocking backwards incompatibility.

Release note: None

ajwerner committed:
sql/schemachanger: add back TXN_DROPPED, stop using OFFLINE

We added the `OFFLINE` and performed write operations during drops to move the
descriptor into that state in order to deal with the fact that synthetically
dropped descriptors would not appear so in virtual tables. Fortunately, this
shortcoming has been rectified. We now revive the TXN_DROPPED state. This
proves to be important because the `OFFLINE` state is not backwards compatible,
and the `TXN_DROPPED` state is. Ideally we'd find a way to move the descriptors
straight to `DROPPED` in the pre-commit case for all the cases we today
support, but doing that would require an overhaul of the dependency rules
which,at this point in the cycle, feels risky.

Fixes #86626

Release justification: Fixes a release-blocking backwards incompatibility.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/86639)
<!-- Reviewable:end -->

ajwerner commented:
This turns out to be insufficient; we need to make it such that the bulk of the work in the drop cases does happen during pre-commit. Otherwise, we're stuck with incompatibilities.

ajwerner committed:
sql/schemachanger/scplan/internal/scstage: accept non-revertible in PreCommit

We can schedule non-revertible operations into PreCommit so long as the next
stage is PostCommitNonRevertible.

Release justification: part of bug fix

Release note: None

ajwerner committed:
sql/schemachanger/scplan/internal/scstage: implicitly satisfy no-op edges

Release note: None

ajwerner committed:
sql/schemachanger/rel: internally store rules with random vars

This is just to avoid bugs later.

Release justification: cleanup

Release note: None

ajwerner committed:
sql/schemachanger/rel: support not-join subqueries

This patch endows rel with the ability to define not-join rules which allow
extra variables to be introduced inside the clauses. When a not-join rule is
invoked, it leads to the outer query finding a contradiction if the rule fully
unifies.

Fixes #77148

Release justification: part of a broader bug-fix

Release note: None

ajwerner committed:
sql/schemachanger: enforce online schema change invariants with dep rules

Previously we had implicit rules in opgen which were used to enforce
sequencing between transitions. This was brittle and made it hard to
collapse certain edges. Now that we can represent a not-join, we can
add the rules we want selectively, only when descriptors are not being
dropped.

Fixes #86691
Fixes #86626

Release justification: fixes a bug

Release note: None

Xiang-Gu commented:
I fixed a few build errors,  rebase on Master, and ran `--rewrite` on affected tests in the above push.

CI now reports test failures; we will need to look into them next.

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/helpers.go:
```diff
@@ -465,3 +465,20 @@ func registerDepRuleForDrop(
+			descriptor.typeFilter(IsDescriptor),
+			descriptor.joinTarget(),
+			joinOnDescIDUntyped(descriptor.el, element, "id"),
+			descriptor.targetStatus(scpb.ToAbsent),
```
should this line be `descriptor.targetStatus(scpb.ToPublic)`?

ajwerner replied:
No, we want this rule to unify if we *are* dropping the descriptor. The `NotJoin` implies negation where it is used if everything inside the rule unifies.

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/dep_drop_object.go:
```diff
@@ -26,8 +26,36 @@ import (
+				from.typeFilter(IsDescriptor),
+				from.el.AttrEqVar(screl.DescID, "_"),
+				from.el.AttrEqVar(rel.Self, to.el),
+				to.el.AttrEqVar(rel.Self, to.el),
```
Isn't this line always true and hence redundant?

ajwerner replied:
Yes, removed

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/dep_drop_object.go:
```diff
@@ -26,8 +26,36 @@ import (
+	registerDepRule(
+		"descriptor txn_drop before descriptor drop",
+		scgraph.PreviousStagePrecedence,
+		"descriptor", "dependent",
```
If this rule is about making sure one descriptor reached `TXN_DROPPED` before reaching `DROPPED` in a previous stage, then we probably should change the name of the second variable `dependent` to, maybe, `descriptor_self`.

ajwerner replied:
Good catch, fixed

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/rules/dep_two_version.go:
```diff
@@ -0,0 +1,171 @@
+	type statusFunc = func(
+		from nodeVars, fromStatus scpb.Status, to nodeVars, toStatus scpb.Status,
+	) rel.Clause
+	addRules := func(c depElement, targetStatus scpb.TargetStatus, sf statusFunc) {
```
it seems `sf statusFunc` is not used. Did I miss something?

Xiang-Gu commented on pkg/sql/schemachanger/scplan/internal/scgraph/edge.go:
```diff
@@ -102,6 +95,10 @@ const (
 	// stage.
 	PreviousStagePrecedence
+
+	// PreviousTransactionPrecedence is like PreviousStagePrecedence but does
+	// not allow the transition to occur if the current phase is not PostCommit.
```
Can you be elobarate a bit more on the comment? Does it mean`from` must be reached before `to` in a previous stage, and both stages must be in PostCommit phase? If so, what do you think of the name `PreviousStageInPostCommitPrecendence`?

postamar replied:
`does not allow the transition to occur unless the current phase is at least PostCommitPhase, because StatementPhase and PreCommitPhase are special in that they take place in the same transaction`?

ajwerner reviewed (commented):
Thanks for taking a look and fixing the issues. The validation things are interesting. I'll get to the bottom of them one way or another. I think I addressed the basic comments, but did it on the old version of the branch. I'll find and add back your changes.

postamar commented on pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go:
```diff
@@ -560,9 +560,18 @@ func (s *TestState) MustReadDescriptor(ctx context.Context, id descpb.ID) catalo
 
+// AddSyntheticDescriptor is part of the scexec.Catalog interface.
+func (s *TestState) AddSyntheticDescriptor(desc catalog.MutableDescriptor) {
+	s.LogSideEffectf("added synthetic descriptor: %v", desc)
```
Can we instead print the yaml diff between the synthetic descriptor and the existing immutable descriptor if it exists?

postamar commented on pkg/sql/schemachanger/scexec/scmutationexec/dependencies.go:
```diff
@@ -117,4 +117,6 @@ type MutationVisitorStateUpdater interface {
 	// RefreshStats refresh stats for a given descriptor.
 	RefreshStats(id descpb.ID)
+	MarkDescriptorSynthetic(id descpb.ID)
+	//	MarkDescriptorSynthetic(id descpb.ID)
```
detritus

postamar commented on pkg/sql/schemachanger/scplan/internal/scstage/build.go:
```diff
@@ -322,7 +322,7 @@ func (sb stageBuilder) isOutgoingOpEdgeAllowed(e *scgraph.OpEdge) bool {
 		return false
 	}
-	if !e.IsPhaseSatisfied(sb.bs.phase) {
+	if !e.IsPhaseSatisfied(sb.bs.phase) && !sb.bc.g.IsNoOp(e) {
```
Nice. By the way, you probably should remove `hasOpEdgeWithOps` and the logic around it in this PR. Search for `// TODO(postamar): uphold the 2-version invariant using dep rules instead.`

postamar commented on pkg/sql/schemachanger/rel/schema_rules.go:
```diff
@@ -87,19 +100,29 @@ var (
+func makeRandomVars(n int) (ret []Var) {
+	ret = make([]Var, n)
+	for i := range ret {
+		ret[i] = Var(hex.EncodeToString(uuid.MakeV4().GetBytes()))
```
This is going to hurt legibility when debugging. Can't we instead use a reserved prefix followed by a counter? You probably can assert that this reserved prefix isn't used in vars outside of rel.

postamar commented on pkg/sql/schemachanger/rel/query_build.go:
```diff
@@ -110,6 +121,16 @@ func (p *queryBuilder) processClause(t Clause) {
+	// if we have a not-join, then we need to find the slots for the inputs
+	// and we have to build the sub-query, which is a whole new query, and
+	// we have to then figure out its depth. In general, its depth should be
+	// the current join depth, so the current
```
incomplete comment

postamar commented on pkg/sql/schemachanger/rel/query_lang_clauses.go:
```diff
@@ -38,7 +38,9 @@ func expanded(c Clauses) Clauses {
-				return !cl.rule.isNotJoin
+				if !cl.rule.isNotJoin {
+					return true
+				}
```
this should instead be in the penultimate commit

ajwerner replied:
Fixed.

postamar commented on pkg/sql/schemachanger/scop/ops.go:
```diff
@@ -30,6 +30,13 @@ const (
 	// performed using internal queries.
 	ValidationType
+
+	maxType
```
This isn't a max, though, it's the max + 1...

In these cases where the enum is unlikely to change in the future it's better to have `const maxType = ValidationType` instead. The alternative (for enums which often have values added to them) is to define a `$<whateverEnumType>ValuesCount` along with `const max$<whateverEnumType> = $<whateverEnumType>ValuesCount - 1`.

ajwerner replied:
meh I fixed the max definition

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/helpers.go:
```diff
@@ -465,3 +469,20 @@ func registerDepRuleForDrop(
+			descriptor.targetStatus(scpb.ToAbsent),
+		}
+	},
+)
```
Very nice.

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/dep_two_version.go:
```diff
@@ -0,0 +1,163 @@
+			panic(err)
+		}
+	}
+	_ = screl.ForEachElementType(func(el scpb.Element) error {
```
We already have a function `forEachElement` in `rules` just like this one.

postamar commented on pkg/sql/schemachanger/scplan/internal/rules/dep_two_version.go:
```diff
@@ -0,0 +1,163 @@
+
+}
+
+// notJoinOnNodeWithStatusIn is a cache to memoize getNotJoinOnNodeWithStatusIn. a not-join rule which will
```
incomplete comment

postamar commented on pkg/sql/schemachanger/scplan/internal/scgraph/edge.go:
```diff
@@ -67,11 +65,6 @@ func (oe *OpEdge) Type() scop.Type {
-// IsPhaseSatisfied returns true iff the operations can run in the given phase.
-func (oe *OpEdge) IsPhaseSatisfied(phase scop.Phase) bool {
-	return phase >= oe.minPhase
-}
```
Good riddance.

postamar commented on pkg/sql/schemachanger/scplan/internal/scstage/build.go:
```diff
@@ -361,7 +358,7 @@ func (sb stageBuilder) nextTargetState(t currentTargetState) currentTargetState
 			next.hasOpEdgeWithOps = true
 		}
-	}
+	}*/
```
Just get rid of this, as suggested in a previous comment, along with `hasOpEdgeWithOps`.

postamar commented on pkg/sql/schemachanger/scplan/internal/scstage/build.go:
```diff
@@ -373,8 +370,8 @@ func (sb stageBuilder) nextTargetState(t currentTargetState) currentTargetState
-		if sb.isUnmetInboundDep(de) {
-			ret = true
+		ret = sb.isUnmetInboundDep(de)
+		if ret {
```
nit: is this really more readable?

postamar commented on pkg/sql/schemachanger/scplan/internal/scstage/build.go:
```diff
@@ -386,6 +383,13 @@ func (sb *stageBuilder) isUnmetInboundDep(de *scgraph.DepEdge) bool {
+		return !fromIsFulfilled ||
+			(sb.bs.phase <= scop.PreCommitPhase &&
+				// If it has been fulfilled implicitly because it's the initial
+				// status, then it doesn't matter the current phase, let it go.
```
`then the current phase doesn't matter`?

postamar reviewed (commented):
This is coming along nicely!

postamar reviewed (approved):
My comments are just nits, so LGTM.

ajwerner reviewed (commented):
I took your changes and addressed the comments.

postamar commented:
CI is failing due to a compilation error:
`pkg/sql/schemachanger/rel/query_build.go:323:3: unnecessary assignment to the blank identifier (S1005)`

Let's 🚢 🇮🇹. My very minor concern about the var generator can wait another day.

ajwerner commented:
TFTR!

bors r=postamar
